pilot/pkg/model/push_context.go

```diff
@@ -1073,6 +1085,17 @@ func (ps *PushContext) ServiceForHostname(proxy *Proxy, hostname host.Name) *Ser
 	return nil
 }
 
+// GatewayServiceForHostname returns the service associated with a given hostname following GW's exported VS.
+func (ps *PushContext) GatewayServiceForHostname(proxy *Proxy, hostname host.Name, patches *MergedEnvoyFilterWrapper) *Service {
+	services := ps.GatewayServices(proxy, patches)
+	for _, svc := range services {
```

Comment:
Seems like this is pretty expensive, will be O(N^2) lookup across services (since we call this for each host).

Probably an easy fix though, we can just compute the Gateway services one time and use between each call?